book/src/json-output.md

```diff
@@ -0,0 +1,157 @@
+
+```console
+$ DEFMT_LOG=debug cargo rb levels > levels.json
+
+(HOST) INFO  flashing program (2 pages / 8.00 KiB)
+└─ probe_run @ src/main.rs:93
+(HOST) INFO  success!
+└─ probe_run @ src/main.rs:126
+────────────────────────────────────────────────────────────────────────────────
+────────────────────────────────────────────────────────────────────────────────
+(HOST) INFO  device halted without error
+└─ probe_run::backtrace @ src/backtrace/mod.rs:108
+```
+
+Afterwards `levels.json` looks like this:
+```json
+{"schema_version":1}
+{"data":"info","host_timestamp":1643113389707243978,"level":"INFO","location":{"file":"src/bin/levels.rs","line":10,"module_path":{"crate_name":"levels","modules":[],"function":"__cortex_m_rt_main"}},"target_timestamp":"0"}
+{"data":"warn","host_timestamp":1643113389707290115,"level":"WARN","location":{"file":"src/bin/levels.rs","line":12,"module_path":{"crate_name":"levels","modules":[],"function":"__cortex_m_rt_main"}},"target_timestamp":"1"}
+{"data":"debug","host_timestamp":1643113389707299759,"level":"DEBUG","location":{"file":"src/bin/levels.rs","line":13,"module_path":{"crate_name":"levels","modules":[],"function":"__cortex_m_rt_main"}},"target_timestamp":"2"}
+{"data":"error","host_timestamp":1643113389707306961,"level":"ERROR","location":{"file":"src/bin/levels.rs","line":14,"module_path":{"crate_name":"levels","modules":[],"function":"__cortex_m_rt_main"}},"target_timestamp":"3"}
+{"data":"println","host_timestamp":1643113389707313290,"level":null,"location":{"file":"src/bin/levels.rs","line":15,"module_path":{"crate_name":"levels","modules":[],"function":"__cortex_m_rt_main"}},"target_timestamp":"4"}
+```
+
+🤔: That seems convenient, but what is this schema version in the first line?
+
+It indicates the version of the json format you are using. `probe-run` will always output it as a header at the beginning of each stream of logs. We anticipate that the format will slightly change while `probe-run` and `defmt` evolve. Using this version you always know which revision is in use and can act upon that.
+
+🤗: Thank you for the explanation!
+
+## JsonFrame
+
+The [`defmt-json-schema`] crate provides a data transfer object `struct JsonFrame` which is used to serialize the log frame to json and can be used to deserialize it back to rust as well.
+
+The `JsonFrame` is defined like this ...
+
+```rust
+pub struct JsonFrame {
+    pub data: String,
+    pub host_timestamp: i64,
+    pub level: Option<log::Level>,
+    pub location: Location,
+    pub target_timestamp: String,
+}
+
+pub struct Location {
+    pub file: Option<String>,
+    pub line: Option<u32>,
+    pub module_path: Option<ModulePath>,
+}
+
+pub struct ModulePath {
+    pub crate_name: String,
+    pub modules: Vec<String>,
+    pub function: String,
+}
+
+# mod log { pub struct Level; }
+```
+
+... which results in following json schema ...
+
+```json
```

Comment:
This is JSON Schema, not JSON. The standard Github JSON renderer makes a mess of it.

Reply:
It does look bad on github, but when rendered in the mdbook it looks good I'd say.

Reply:
OK, well I guess most people are looking at the MdBook.